### DISCUSSION TITLE
Tycho 4.x - Compiler error "The method asList(T...) of Arrays.."

### DISCUSSION THREAD

**Discussion by Waschndolos** [eclipse-tycho/tycho#2931] [Q&A] [answered]
Quick question in case anybody knows. I'm trying to update a huge project to Tycho 4.0.3. For one of our test plugins I get the following compile error:

```
[ERROR]         checkMultiValues(Arrays.asList("1", Double.valueOf(50.0), Integer.valueOf(150), "254"), Arrays.asList("2", Integer.valueOf(52), Double.valueOf(149.0)), false, SomeEnum.FLOAT32);
[ERROR]                                 ^^^^^^
[ERROR] The method asList(T...) of type Arrays is not applicable as the formal varargs element type T is not accessible here
```
The Manifest.MF of the plugin contains: 

```
Manifest-Version: 1.0
Bundle-RequiredExecutionEnvironment: JavaSE-11
Bundle-ManifestVersion: 2
Bundle-SymbolicName: my.name
Bundle-Version: 99.99.99
Eclipse-BundleShape: dir
Bundle-Name: My Name
Bundle-Vendor: My Company
Require-Bundle: some.other.dependency,
 another.dependency,
Bundle-ActivationPolicy: lazy
```

I'm using the BREE configuration in the tycho-compiler plugin: 
```
	<plugin>
		<groupId>org.eclipse.tycho</groupId>
		<artifactId>tycho-compiler-plugin</artifactId>
		<version>${tycho.version}</version>
		<configuration>
			<useJDK>BREE</useJDK>
			<compilerArgs>
				<arg>-err:-forbidden</arg>
				<arg>-proc:none</arg>
				<arg>--module-path</arg>
				<arg>${java.home}/jmods</arg>
				<arg>--add-modules</arg>
				<arg>java.smartcardio</arg>
			</compilerArgs>
		</configuration>
	</plugin>
```

I have a toolchains.xml with <id>JavaSE-11</id> and <id>JavaSE-1.8</id>. The Runtime for Maven 3.9.5 is Java 17.0.5. 

The same xml configuration works fine with tycho 2.0.5. I assume the error is rooted in the jdt compiler somewhere which tycho 4.x uses (using Eclipse Compiler for Java(TM) 3.35.0.v20230814-2020). 

Does anyone have an idea where I can look? Struggling with this for quite some time now. Maybe I'm missing a compiler flag?

**Comment by laeubi** [comment]
You can try to use an older ECJ version
```

<plugin>
  <groupId>org.eclipse.tycho</groupId>
  <artifactId>tycho-compiler-plugin</artifactId>
  <version>${tycho-version}</version>
  <dependencies>
    <dependency>
      <groupId>org.eclipse.jdt</groupId>
      <artifactId>ecj</artifactId>
      <version>....</version>
    </dependency>
  </dependencies>
</plugin>
```

If you identified the compiler version where the issue started to show up, it would help to ask at the JDT project for further guidance:

https://github.com/eclipse-jdt/eclipse.jdt.core

**Reply by Waschndolos** [reply]
Ah ok perfect. I'll travel back in time to see where it happened :). Keeping the discussion open until I found it.

**Reply by Waschndolos** [reply]
Hm strange I'm down to  3.28.0.v20211117-1416 now but it still occurs

**Comment by Waschndolos** [comment, marked as answer]
Ok I found the (stupid) error. In the root pom we had:

```
<plugin>
  <groupId>org.eclipse.tycho</groupId>
  <artifactId>tycho-compiler-plugin</artifactId>
  <version>${tycho.version}</version>
  <configuration>
	  <useJDK>BREE</useJDK>
	  <compilerArgs>
		  <arg>-err:-forbidden</arg>
		  <arg>-proc:none</arg>
		  <arg>--module-path</arg>
		  <arg>${java.home}/jmods</arg>
		  <arg>--add-modules</arg>
		  <arg>java.smartcardio</arg>
	  </compilerArgs>
	  <deriveReleaseCompilerArgumentFromTargetLevel>false</deriveReleaseCompilerArgumentFromTargetLevel>
  </configuration>
</plugin>
```

Let's all guess what `${java.home}` is when I run the maven build with Java 17 Runtime -.- So we tried to compile with Java 11 from the toolchains.xml together with the jmods of Java 17. 

So removing 
```
<arg>--module-path</arg>
<arg>${java.home}/jmods</arg>
```
fixed it.
